### PR TITLE
New: implements PSR-4 Autoloading Standard for Improved Code Structure and Maintainability

### LINKED ISSUE #563: Migrate to PSR-4 autoload and ditch classmap
Migrate to PSR-4 autoload and ditch classmap, which will require files and directories to be case sensitive and the corresponding classes and namespaces to be in PascalCase.

### PR DESCRIPTION
**Pull request by alexmigf**
closes #563 

- Refactored files to follow PSR-4 autoloading standards
- Updated namespaces to better match directory structure
- Modified composer.json for PSR-4 autoloading
- Ensured backward compatibility for legacy classes and namespaces
- Renamed functions file and moved to plugin root directory
- Configured Strauss to prefix vendor namespaces

### Note:

Strauss doesn't correctly apply the new namespace to the `FontLib` class variables, so a manual fix is needed to ensure proper font subsetting functionality. Fix applied here: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/pull/840/commits/4643b01e130b52797dce65acce2257532dfd3b0b

### CODE REVIEW

**Review by dwalkerpriv — APPROVED** [review]
Normal functions seem to work fine 👍 . I tested:

- All document generation
- Document uploads
- Auto and manual creation
- Document attachments

Didn't notice any slow downs, nor any errors in the logs

**Review by MohamadNateqi — CHANGES_REQUESTED** [review]

File: includes/Makers/PdfMaker.php [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
 
 if ( ! defined( 'ABSPATH' ) ) {
 	exit; // Exit if accessed directly
 }
 
-if ( ! class_exists( '\\WPO\\WC\\PDF_Invoices\\Makers\\PDF_Maker' ) ) :
+if ( ! class_exists( '\\WPO\\IPS\\Makers\\PdfMaker' ) ) :
 
-class PDF_Maker {
+class PdfMaker {
```
[the remark this answers is not in the document]
Comment (alexmigf) [reply includes/Makers/PdfMaker.php]: Was just to follow the same pattern as the others, but I have changed now.

File: composer.json [hunk 2]
```diff
@@ -1,19 +1,57 @@
...
+			"override_autoload": {
+			},
+			"exclude_from_copy": {
+				"packages": [
+				],
+				"namespaces": [
+				],
+				"file_patterns": [
+				]
+			},
```
[the remark this answers is not in the document]
Comment (alexmigf) [reply composer.json:37]: They can, but I think it helps if you want to make adjustments in the future.

**Comment by BrunoPavlinic98** [comment]
@alexmigf I received the following fatal error when trying to update the order from "Pending payment" status to "Processing" status. I did that because I wanted to test email attachments, and I have set the **Invoice** **Attach to:** option to be **Processing order** and **Completed order**. Here is the error from log:

[30-Jul-2024 08:55:15 UTC] PHP Fatal error:  Cannot declare class WPO\IPS\Vendor\Sabberworm\CSS\CSSList\CSSBlockList, because the name is already in use in C:\Users\Bruno Pavlinic\Desktop\WPOvernight\Sites\wpovernight\app\public\wp-content\plugins\woocommerce-pdf-invoices-packing-slips\vendor\sabberworm\php-css-parser\src\CSSList\CSSBlockList.php on line 24

I think this is related to this PR.
I was using PDF IPS PRO main branch and Templates main branch.

**Comment by alexmigf** [comment]
@BrunoPavlinic98 could you try my last commit?

File: ubl/Handlers/Ubl/IdHandler.php [hunk 3]
```diff
@@ -1,9 +1,9 @@
 <?php
 
-namespace WPO\WC\UBL\Handlers\Ubl;
+namespace WPO\IPS\UBL\Handlers\Ubl;
 
-use WPO\WC\UBL\Handlers\UblHandler;
-use WPO\WC\UBL\Models\Order;
+use WPO\IPS\UBL\Handlers\UblHandler;
+use WPO\IPS\UBL\Models\Order;
```
[the remark this answers is not in the document]
Comment (alexmigf) [reply ubl/Handlers/Ubl/IdHandler.php]: Done.

**Comment by BrunoPavlinic98** [comment]
> @BrunoPavlinic98 could you try my last commit?

@alexmigf Your commit solved the issue.

**Comment by alexmigf** [comment]
@BrunoPavlinic98 great :100:

**Comment by dpeyou** [comment]
I am getting this vendor file error:

```

Fatal error: Class "FontLib\TrueType\File" not found

/home/dtoo/Local Sites/test/app/public/wp-content/plugins/woocommerce-pdf-invoices-packing-slips/vendor/dompdf/php-font-lib/src/FontLib/Font.php (67)

#0 /home/dtoo/Local Sites/test/app/public/wp-content/plugins/woocommerce-pdf-invoices-packing-slips/vendor/dompdf/dompdf/lib/Cpdf.php(1180): WPO\IPS\Vendor\FontLib\Font::load('/home/dtoo/Loca...')
#1 /home/dtoo/Local Sites/test/app/public/wp-content/plugins/woocommerce-pdf-invoices-packing-slips/vendor/dompdf/dompdf/lib/Cpdf.php(944): WPO\IPS\Vendor\Dompdf\Cpdf->processFont(8, Array)
#2 /home/dtoo/Local Sites/test/app/public/wp-content/plugins/woocommerce-pdf-invoices-packing-slips/vendor/dompdf/dompdf/lib/Cpdf.php(3208): WPO\IPS\Vendor\Dompdf\Cpdf->o_font(8, 'add')
#3 /home/dtoo/Local Sites/test/app/public/wp-content/plugins/woocommerce-pdf-invoices-packing-slips/vendor/dompdf/dompdf/src/Adapter/CPDF.php(990): WPO\IPS\Vendor\Dompdf\Cpdf->output(false)
#4 /home/dtoo/Local Sites/test/app/public/wp-content/plugins/woocommerce-pdf-invoices-packing-slips/vendor/dompdf/dompdf/src/Dompdf.php(935): WPO\IPS\Vendor\Dompdf\Adapter\CPDF->output(Array)
#5 /home/dtoo/Local Sites/test/app/public/wp-content/plugins/woocommerce-pdf-invoices-packing-slips/includes/Makers/PDFMaker.php(57): WPO\IPS\Vendor\Dompdf\Dompdf->output()
#6 /home/dtoo/Local Sites/test/app/public/wp-content/plugins/woocommerce-pdf-invoices-packing-slips/includes/Documents/OrderDocument.php(1027): WPO\IPS\Makers\PDFMaker->output()
#7 /home/dtoo/Local Sites/test/app/public/wp-content/plugins/woocommerce-pdf-invoices-packing-slips/includes/Documents/OrderDocument.php(1093): WPO\IPS\Documents\OrderDocument->get_pdf()
#8 /home/dtoo/Local Sites/test/app/public/wp-content/plugins/woocommerce-pdf-invoices-packing-slips/includes/Main.php(569): WPO\IPS\Documents\OrderDocument->output_pdf('inline')
#9 /home/dtoo/Local Sites/test/app/public/wp-includes/class-wp-hook.php(324): WPO\IPS\Main->generate_document_ajax('')
#10 /home/dtoo/Local Sites/test/app/public/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters('', Array)
#11 /home/dtoo/Local Sites/test/app/public/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#12 /home/dtoo/Local Sites/test/app/public/wp-admin/admin-ajax.php(192): do_action('wp_ajax_generat...')
#13 {main}


```

Trying to do a composer update leaves me with this error below. Problem 1 mentions **dompdf**


```

$ composer update 
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Root composer.json requires dompdf/dompdf ^3.0 -> satisfiable by dompdf/dompdf[v3.0.0].
    - dompdf/dompdf v3.0.0 requires ext-dom * -> it is missing from your system. Install or enable PHP's dom extension.
  Problem 2
    - sabre/xml[2.2.5, ..., 2.2.9] require ext-xmlwriter * -> it is missing from your system. Install or enable PHP's xmlwriter extension.
    - Root composer.json requires sabre/xml ^2.2.5 -> satisfiable by sabre/xml[2.2.5, ..., 2.2.9].

To enable extensions, verify that they are enabled in your .ini files:
    - /etc/php/7.4/cli/php.ini
    - /etc/php/7.4/cli/conf.d/10-opcache.ini
    - /etc/php/7.4/cli/conf.d/10-pdo.ini
    - /etc/php/7.4/cli/conf.d/20-calendar.ini
    - /etc/php/7.4/cli/conf.d/20-ctype.ini
    - /etc/php/7.4/cli/conf.d/20-curl.ini
    - /etc/php/7.4/cli/conf.d/20-exif.ini
    - /etc/php/7.4/cli/conf.d/20-ffi.ini
    - /etc/php/7.4/cli/conf.d/20-fileinfo.ini
    - /etc/php/7.4/cli/conf.d/20-ftp.ini
    - /etc/php/7.4/cli/conf.d/20-gd.ini
    - /etc/php/7.4/cli/conf.d/20-gettext.ini
    - /etc/php/7.4/cli/conf.d/20-iconv.ini
    - /etc/php/7.4/cli/conf.d/20-json.ini
    - /etc/php/7.4/cli/conf.d/20-mbstring.ini
    - /etc/php/7.4/cli/conf.d/20-phar.ini
    - /etc/php/7.4/cli/conf.d/20-posix.ini
    - /etc/php/7.4/cli/conf.d/20-readline.ini
    - /etc/php/7.4/cli/conf.d/20-shmop.ini
    - /etc/php/7.4/cli/conf.d/20-sockets.ini
    - /etc/php/7.4/cli/conf.d/20-sysvmsg.ini
    - /etc/php/7.4/cli/conf.d/20-sysvsem.ini
    - /etc/php/7.4/cli/conf.d/20-sysvshm.ini
    - /etc/php/7.4/cli/conf.d/20-tokenizer.ini
You can also run `php --ini` inside terminal to see which files are used by PHP in CLI mode.


```

**Comment by alexmigf** [comment]
@dpeyou don't run any composer command, we have now Strauss implemented which should handle that. Can you try again without doing that?

**Comment by dpeyou** [comment]
@alexmigf it appears the same error happens after the latest commit :frowning_face: 

``` 

dtoo@dtooHuaMx:~/Local Sites/test/app/public/wp-content/plugins/woocommerce-pdf-invoices-packing-slips
$ git pull
hint: Pulling without specifying how to reconcile divergent branches is
hint: discouraged. You can squelch this message by running one of the following
hint: commands sometime before your next pull:
hint: 
hint:   git config pull.rebase false  # merge (the default strategy)
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint: 
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (5/5), done.
remote: Compressing objects: 100% (1/1), done.
remote: Total 3 (delta 2), reused 3 (delta 2), pack-reused 0
Unpacking objects: 100% (3/3), 333 bytes | 111.00 KiB/s, done.
From https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips
   6c26b265..e3de861c  psr4-strauss-v2 -> origin/psr4-strauss-v2
Updating 6c26b265..e3de861c
Fast-forward
 .distignore | 3 +++
 1 file changed, 3 insertions(+)


```

**Comment by alexmigf** [comment]
@dpeyou you must have something wrong with your vendor folder. The error points to a class that doesn't exist, and in my case I don't have that namespace anywhere. Also the line doesn't seem correct to me, so I think your vendor directory has wrong versions of the library. I think you need to delete the plugin and do again.

**Comment by dpeyou** [comment]
@alexmigf sadly I am getting the same error, it seems.

I deleted the plugin, did `git clone`, then `git checkout`.

I'm not sure if a staging site will help, but I am pushing the site to a server you'll be able to check out. 
(Credentials via Slack).

**Comment by alexmigf** [comment]
@dpeyou that would help yes.

**Review by dwalkerpriv — APPROVED** [review]
Re-ran my tests all seem to work for me. I wasn't getting the error Bruno had and don't it get with the recent updates 👍

**Comment by alexmigf** [comment]
@dpeyou should be fixed now, pull the latest changes.

**Review by dpeyou — APPROVED** [review]
LGTM now, I cannot find any errors :rocket:

**Comment by dwalkerpriv** [comment]
I've been using this since Monday I believe. Havent had any issues with the normal functions of plugin nor with the compatibility of the PT or Pro

**Review by BrunoPavlinic98 — APPROVED** [review]
LGTM!
The code looks OK.
I have tested various different IPS functionalities without any issues. I will keep my free plugin branch on this one for some time just to test this while working on another tasks.

**Review by YordanSoares — APPROVED** [review]
I've been testing this PR since last Wednesday, after the weekly meeting, and haven't gotten any issues so far. I have tested this with Professional, Premium Templates and, mPDF, and everything seems to be working correctly.

Within the testing, I took advantage to check a reported conflict with dompdf caused by “TI WooCommerce Wishlist Premium” (see the issue reports [here](https://secure.helpscout.net/search/?query=%22TI%20WooCommerce%20Wishlist%20Premium%22%20tag%3Adompdf)), and I can confirm that this is solved in this PR.

**Review by MohamadNateqi — APPROVED** [review]
Works great!

**Review by Terminator-Barbapapa — APPROVED** [review]
🚀

**Comment by alexmigf** [comment]
@szepeviktor we are running Strauss; Composer is handled by it.
